packages/client/src/features/smart-connector/smart-connector.ts

```diff
@@ -0,0 +1,524 @@
+        if (!this.editorContext.isReadonly) {
+            item.actions.forEach(e => {
+                let args: Args;
+                if (TriggerEdgeCreationAction.is(e)) {
+                    args = { source: this.selectedElementId };
+                    (e as TriggerEdgeCreationAction).args = args;
+                }
+                if (TriggerNodeCreationAction.is(e)) {
+                    args = { createEdge: true, source: this.selectedElementId, edgeType: (item as SmartConnectorNodeItem).edgeType };
+                    (e as TriggerNodeCreationAction).args = args;
+                }
+            });
+            this.actionDispatcher.dispatchAll(item.actions.concat([SetUIExtensionVisibilityAction.create(
+                { extensionId: SmartConnector.ID, visible: false })
+            ]));
+            this.hideAll();
+        }
+    }
+
+    // #endregion
+
+    handle(action: Action): ICommand | Action | void {
+        if (OpenSmartConnectorAction.is(action)) {
+            this.selectedElementId = action.selectedElementID;
+            this.actionDispatcher.dispatch(
+                SetUIExtensionVisibilityAction.create({ extensionId: SmartConnector.ID, visible: true })
+            );
+        }
+        else {
+            this.actionDispatcher.dispatch(
+                SetUIExtensionVisibilityAction.create({ extensionId: SmartConnector.ID, visible: false })
+            );
+            this.hideAll();
+        }
+    }
+
+    async preRequestModel(): Promise<void> {
+        const requestAction = RequestContextActions.create({
+            contextId: SmartConnector.ID,
+            editorContext: {
+                selectedElementIds: []
+            }
+        });
+        const response = await this.actionDispatcher.request<SetContextActions>(requestAction);
+        this.smartConnectorItems = response.actions.map(e => e as SmartConnectorGroupItem);
+    }
+}
+
+export function showSmartConnector(container: HTMLElement, expandButton: HTMLElement): void {
+    container.style.visibility = 'visible';
+    expandButton.style.visibility = 'hidden';
+}
+
+function getHeaderIfGroupContainsCollapsable(group: HTMLElement): HTMLElement | undefined {
+    if (group && group.getElementsByClassName(COLLAPSABLE_CLASS).length !== 0) {
+        return group.getElementsByClassName(HEADER_CLASS)[0] as HTMLElement;
+    }
+    return undefined;
+}
+
+@injectable()
+export class SmartConnectorKeyListener extends KeyListener {
+    override keyDown(_element: GModelElement, event: KeyboardEvent): Action[] {
+        const container = document.getElementById(SMART_CONNECTOR_CONTAINER_ID)!;
+        const expandButton = document.getElementById(EXPAND_BUTTON_ID);
+        const smartConnector = expandButton?.parentElement;
+        if (matchesKeystroke(event, 'Space') && smartConnector?.style.visibility === 'visible'
+            && expandButton?.style.visibility === 'visible') {
+            showSmartConnector(container, expandButton);
```

Comment:
I'm not sure if it is okay how the extension is expanded. You are currently hiding it with CSS and then make it visible again with CSS.

I think the whole state change (circle -> expanded) should be handled in the class directly. I would say, create a new `Action` with the state you want e.g., collapsed, expanded as a parameter and trigger it here. Then let your `UIExtension` handle the action and depending on the provided `State` update it to be either `collapsed` or `expanded`.

So if you click space, you would tell your `UIExtension` that it should move from the circle view to the box view.

Reply:
You're right, I am not sure why I initially did it this way.

I added the new Action now, but I did not add a key binding to close the Smart Connector with (since I am not sure which keys I can use so they don't overlap), just Space to open it.